sw/rust/ibex-demo-system-hal/src/gpio/dyn_pin.rs

```diff
@@ -0,0 +1,98 @@
+// Copyright lowRISC contributors.
+// Licensed under the Apache License, Version 2.0, see LICENSE for details.
+// SPDX-License-Identifier: Apache-2.0
+
+use super::pin::*;
+use embedded_hal::digital::v2::{InputPin, OutputPin, PinState};
+
+#[derive(PartialEq)]
+enum PinConfig {
+    Output,
+    Input,
+}
+
+pub struct DynPin {
+    port: char,
+    id: u8,
+    state: PinConfig,
+}
+
+impl DynPin {
+    pub fn new(port: char, id: u8) -> Self {
+        Self {
+            port,
+            id,
+            state: PinConfig::Input,
+        }
+    }
+}
+
+impl Pin for DynPin {
+    fn id(&self) -> u8 {
+        self.id
+    }
+
+    fn port(&self) -> char {
+        self.port
+    }
+}
+
+impl DynPin {
+    pub fn into_input(mut self) -> Result<DynPin, crate::utils::Error> {
+        self.state = PinConfig::Input;
```

Comment:
What's the reason for not using type state for input/output in `DynPin`?

Reply:
I was overseeing a case where an application would want a pin to be input or output depending on a user input, in this case the type state wouldn't work right?
Other reason is to have the pins with the same type which make it ease to capture them in an array.

Reply:
I don't think you want to have pins of different states in an array, though. Even you want, it would be more clear to have `Either<DynPin<Input>, DynPin<Output>>` then just a `DynPin` because it'll be prone to misuse.

Reply:
Changed.